pytest 8 compatibility

### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of scanpy.
- [X] (optional) I have confirmed this bug exists on the master branch of scanpy.

### What happened?

When running the tests with pytest<=8, the doctest for `scanpy.preprocessing._simple.filter_cells` errors in a way I can't quite figure out how to fix. 

I think what's happening is that the "error on warning" isn't being overridden correctly when we expect the test to warn. Possibly related to https://github.com/pytest-dev/pytest/issues/11759

@flying-sheep any ideas how to fix? I will just pin pytest for now.

### Minimal code sample

```python
Examples
    --------
    >>> import scanpy as sc
    >>> adata = sc.datasets.krumsiek11()
    UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
        utils.warn_names_duplicates("obs")
```


### Error output

```pytb
======================================================================================================================= FAILURES =======================================================================================================================
_________________________________________________________________________________________________ [doctest] scanpy.preprocessing._simple.filter_cells __________________________________________________________________________________________________
081         Boolean index mask that does filtering. `True` means that the
082         cell is kept. `False` means the cell is removed.
083     number_per_cell
084         Depending on what was thresholded (`counts` or `genes`),
085         the array stores `n_counts` or `n_cells` per gene.
086 
087     Examples
088     --------
089     >>> import scanpy as sc
090     >>> adata = sc.datasets.krumsiek11()
UNEXPECTED EXCEPTION: UserWarning('Observation names are not unique. To make them unique, call `.obs_names_make_unique`.')
Traceback (most recent call last):
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/doctest.py", line 1353, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest scanpy.preprocessing._simple.filter_cells[1]>", line 1, in <module>
  File "/mnt/workspace/repos/scanpy/scanpy/datasets/_datasets.py", line 109, in krumsiek11
    adata = read(filename, first_column_names=True)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/legacy_api_wrap/__init__.py", line 80, in fn_compatible
    return fn(*args_all, **kw)
           ^^^^^^^^^^^^^^^^^^^
  File "/mnt/workspace/repos/scanpy/scanpy/readwrite.py", line 124, in read
    return _read(
           ^^^^^^
  File "/mnt/workspace/repos/scanpy/scanpy/readwrite.py", line 801, in _read
    adata = read_text(filename, delimiter, first_column_names)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/anndata/_io/read.py", line 360, in read_text
    return _read_text(f, delimiter, first_column_names, dtype)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/anndata/_io/read.py", line 471, in _read_text
    return AnnData(
           ^^^^^^^^
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/anndata/_core/anndata.py", line 362, in __init__
    self._init_as_actual(
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/anndata/_core/anndata.py", line 574, in _init_as_actual
    self._check_uniqueness()
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/anndata/_core/anndata.py", line 1906, in _check_uniqueness
    utils.warn_names_duplicates("obs")
  File "/mnt/workspace/mambaforge/envs/scanpy-dev/lib/python3.11/site-packages/anndata/utils.py", line 275, in warn_names_duplicates
    warnings.warn(
UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
/mnt/workspace/repos/scanpy/scanpy/preprocessing/_simple.py:90: UnexpectedException
```


### Versions

<details>

```
-----
anndata     0.10.5.post1
scanpy      1.10.0.dev197+g96e19540
-----
IPython             8.20.0
PIL                 10.2.0
asciitree           NA
asttokens           NA
cloudpickle         3.0.0
cycler              0.12.1
cython_runtime      NA
dask                2024.1.1
dateutil            2.8.2
decorator           5.1.1
executing           2.0.1
fasteners           0.19
h5py                3.10.0
igraph              0.11.3
iniconfig           NA
jedi                0.19.1
jinja2              3.1.3
joblib              1.3.2
kiwisolver          1.4.5
legacy_api_wrap     NA
leidenalg           0.10.2
llvmlite            0.41.1
markupsafe          2.1.4
matplotlib          3.8.2
mpl_toolkits        NA
msgpack             1.0.7
natsort             8.4.0
numba               0.58.1
numcodecs           0.12.1
numpy               1.26.3
packaging           23.2
pandas              2.2.0
parso               0.8.3
pexpect             4.9.0
pluggy              1.4.0
prompt_toolkit      3.0.43
psutil              5.9.8
ptyprocess          0.7.0
pure_eval           0.2.2
py                  NA
pygments            2.17.2
pyparsing           3.1.1
pytest              8.0.0
pytz                2023.4
scipy               1.12.0
session_info        1.0.0
six                 1.16.0
sklearn             1.4.0
sparse              0.15.1
stack_data          0.6.3
tblib               3.0.0
texttable           1.7.0
threadpoolctl       3.2.0
tlz                 0.12.1
toolz               0.12.1
traitlets           5.14.1
wcwidth             0.2.13
yaml                6.0.1
zarr                2.16.1
zipp                NA
-----
Python 3.11.7 | packaged by conda-forge | (main, Dec 23 2023, 14:43:09) [GCC 12.3.0]
Linux-5.15.0-87-generic-x86_64-with-glibc2.35
-----
Session information updated at 2024-01-29 14:16
```

</details>

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -82,7 +82,7 @@ scanpy = "scanpy.cli:console_main"
 
 [project.optional-dependencies]
 test-min = [
-    "pytest>=7.4.2,<8",
+    "pytest>=7.4.2",
     "pytest-nunit",
     "pytest-mock",
     "profimp",

--- a/scanpy/testing/_doctests.py
+++ b/scanpy/testing/_doctests.py
@@ -24,7 +24,7 @@ def doctest_needs(mod: str) -> Callable[[F], F]:
 
     def decorator(func: F) -> F:
         if mark is not None:
-            func._doctest_mark = mark
+            func._doctest_needs = mark
         return func
 
     return decorator

--- a/scanpy/testing/_pytest/__init__.py
+++ b/scanpy/testing/_pytest/__init__.py
@@ -12,7 +12,7 @@ from .fixtures import *  # noqa: F403
 if TYPE_CHECKING:
     from collections.abc import Generator, Iterable
 
-doctest_env_marker = pytest.mark.usefixtures("doctest_env")
+    from .marks import needs
 
 
 # Defining it here because it’s autouse.
@@ -59,16 +59,19 @@ def pytest_collection_modifyitems(
             item.add_marker(skip_internet)
 
 
-def pytest_itemcollected(item: pytest.Item) -> None:
-    import pytest
-
-    if not isinstance(item, pytest.DoctestItem):
+@pytest.fixture(autouse=True)
+def _modify_doctests(request: pytest.FixtureRequest) -> None:
+    if not isinstance(request.node, pytest.DoctestItem):
         return
 
-    item.add_marker(doctest_env_marker)
+    request.getfixturevalue("doctest_env")
 
-    func = _import_name(item.name)
-    if marker := getattr(func, "_doctest_mark", None):
-        item.add_marker(marker)
-    if skip_reason := getattr(func, "_doctest_skip_reason", False):
-        item.add_marker(pytest.mark.skip(reason=skip_reason))
+    func = _import_name(request.node.name)
+    needs_marker: needs | None
+    if needs_marker := getattr(func, "_doctest_needs", None):
+        assert needs_marker.mark.name == "skipif"
+        if needs_marker.mark.args[0]:
+            pytest.skip(reason=needs_marker.mark.kwargs["reason"])
+    skip_reason: str | None
+    if skip_reason := getattr(func, "_doctest_skip_reason", None):
+        pytest.skip(reason=skip_reason)

--- a/scanpy/testing/_pytest/__init__.py
+++ b/scanpy/testing/_pytest/__init__.py
@@ -19,14 +19,17 @@ if TYPE_CHECKING:
 @pytest.fixture(autouse=True)
 def _global_test_context(request: pytest.FixtureRequest) -> Generator[None, None, None]:
     """Switch to agg backend, reset settings, and close all figures at teardown."""
+    # make sure seaborn is imported and did its thing
+    import seaborn as sns  # noqa: F401
     from matplotlib import pyplot as plt
+    from matplotlib.testing import setup
 
-    from scanpy import settings
+    import scanpy as sc
 
-    plt.switch_backend("agg")
-    settings.logfile = sys.stderr
-    settings.verbosity = "hint"
-    settings.autoshow = True
+    setup()
+    sc.settings.logfile = sys.stderr
+    sc.settings.verbosity = "hint"
+    sc.settings.autoshow = True
 
     if isinstance(request.node, pytest.DoctestItem):
         _modify_doctests(request)

--- a/scanpy/testing/_pytest/__init__.py
+++ b/scanpy/testing/_pytest/__init__.py
@@ -65,7 +65,7 @@ def pytest_collection_modifyitems(
 def _modify_doctests(request: pytest.FixtureRequest) -> None:
     assert isinstance(request.node, pytest.DoctestItem)
 
-    request.getfixturevalue("doctest_env")
+    request.getfixturevalue("_doctest_env")
 
     func = _import_name(request.node.name)
     needs_marker: needs | None

--- a/scanpy/testing/_pytest/__init__.py
+++ b/scanpy/testing/_pytest/__init__.py
@@ -17,7 +17,7 @@ if TYPE_CHECKING:
 
 # Defining it here because it’s autouse.
 @pytest.fixture(autouse=True)
-def global_test_context() -> Generator[None, None, None]:
+def _global_test_context(request: pytest.FixtureRequest) -> Generator[None, None, None]:
     """Switch to agg backend, reset settings, and close all figures at teardown."""
     from matplotlib import pyplot as plt
 
@@ -27,6 +27,9 @@ def global_test_context() -> Generator[None, None, None]:
     settings.logfile = sys.stderr
     settings.verbosity = "hint"
     settings.autoshow = True
+
+    if isinstance(request.node, pytest.DoctestItem):
+        _modify_doctests(request)
 
     yield
 
@@ -59,10 +62,8 @@ def pytest_collection_modifyitems(
             item.add_marker(skip_internet)
 
 
-@pytest.fixture(autouse=True)
 def _modify_doctests(request: pytest.FixtureRequest) -> None:
-    if not isinstance(request.node, pytest.DoctestItem):
-        return
+    assert isinstance(request.node, pytest.DoctestItem)
 
     request.getfixturevalue("doctest_env")
 

--- a/scanpy/testing/_pytest/fixtures/__init__.py
+++ b/scanpy/testing/_pytest/fixtures/__init__.py
@@ -22,7 +22,7 @@ if TYPE_CHECKING:
 
 __all__ = [
     "float_dtype",
-    "doctest_env",
+    "_doctest_env",
     "_pbmc3ks_parametrized_session",
     "pbmc3k_parametrized",
     "pbmc3k_parametrized_small",
@@ -35,7 +35,7 @@ def float_dtype(request):
 
 
 @pytest.fixture()
-def doctest_env(cache: pytest.Cache, tmp_path: Path) -> Generator[None, None, None]:
+def _doctest_env(cache: pytest.Cache, tmp_path: Path) -> Generator[None, None, None]:
     from scanpy import settings
     from scanpy._compat import chdir
 

--- a/scanpy/tests/test_plotting.py
+++ b/scanpy/tests/test_plotting.py
@@ -3,22 +3,17 @@ from __future__ import annotations
 from functools import partial
 from itertools import chain, combinations, repeat
 from pathlib import Path
-
-import pytest
-from matplotlib.testing import setup
-from packaging import version
-
-setup()
-
 from typing import TYPE_CHECKING
 
 import matplotlib as mpl
 import matplotlib.pyplot as plt
 import numpy as np
 import pandas as pd
+import pytest
 import seaborn as sns
 from anndata import AnnData
 from matplotlib.testing.compare import compare_images
+from packaging import version
 
 import scanpy as sc
 from scanpy._compat import pkg_version
@@ -35,9 +30,6 @@ if TYPE_CHECKING:
 
 HERE: Path = Path(__file__).parent
 ROOT = HERE / "_images"
-
-sc.pl.set_rcParams_defaults()
-sc.set_figure_params(dpi=40, color_map="viridis")
 
 
 # Test images are saved in the directory ./_images/<test-name>/

